# Queuing Homework

### 15.11 Road transport company problem.

Given parameters:
\begin{align*}
\lambda &= \frac{8 customer}{hour} \\
\mu &= \frac{12 customer}{hour} \\
L_q &= \frac{\lambda^2}{\mu(\mu-\lambda)} = \frac{8\cdot8}{12\cdot 4} = \frac{16}{3}  =\text{average number customer in queue}\\
W_q &= \frac{\lambda}{\mu(\mu-\lambda)} = \frac{8}{12\cdot4}=\frac16 hour/cust=10 minutes/cust
\end{align*}

We see that the average number of customers in queue is 16/3. We also found that the average time a customer waits before getting service is ten minutes. 

To evaluate whether the new system is worth it, we see that the current cost of goodwill is $480 minutes \cdot \frac{16}3 \cdot10 minutes \cdot\frac{12}{minute}= 7.111 $

We calulate this to be 307200 paise per day which equals 3072 rupees per day.  

In [1]:
480 * (16/3)*12*10

307200.0

Next we calcuate this again at the new $\mu$ value. If service time is three minutes, 

then $\mu = \frac{customer}{3min}\cdot\frac{60min}{hour} = \frac{20 cust}{hour} $

Thus, $L_q = \frac{64}{240} = \frac4{15}$

and $W_q = \frac{8}{20\cdot12} =\frac{8}{240} =\frac1{30} hour/cust = 2min/cust$

The new cost of goodwill with becomes

(minutes per day)$ \cdot L_q \cdot W_q \cdot $cost of goodwill per minute in queue

480 min per day $\cdot \frac4{15} \cdot 2 min/cust \cdot 12$paise per minute

=$ 3072$ paise per day = 30.72 rupees per day. 

In [1]:
480*(4/15)*2*12

3072.0

Thus the money saved by the new server is $307200-3072 = 304128 $ paise = 3041.28 rupees per day. 

Since cost saved of 3041.28 rupees is greater than the added expense of 50 rupees, then it is worth it.

In [2]:
307200-3072

304128

## 15.21  Car servicing station problem. 

First, we decide what the parameters of the model are based on the question. We find that the model is 
__M/M/2/FIFO/4/$\infty$__. We also find, 
\begin{align*}
\lambda &= 12 \text{cars/day} \\
\mu&=8 \text{cars/day/server} \\
\end{align*}

Since we have correctly identified our model, then we can use the formulas for the model with these parameters. To use these formulae, we first need our $p_0$ value. For this model, our $p_0$ is 
$$
p_0 = \frac{1-\rho}{1-\rho^{M+1}}
$$
where $\rho = \frac{\lambda}{c\mu}$ and $M$ is the system capacity. 

In [60]:
#### p_0 calculation. 
import math
lamb = 12
mu=8
k=2 # number of servers
M = 4 # system capacity
rho=(lamb/(k*mu))
p0 = 0.0
denom = 0.0
# prob models page 247
denom = (1+k*(rho))+((k**k)/math.factorial(k))*(rho**k)*((1-(rho**(M-k+1)))/(1-rho))
p0 = 1/denom
print('p0 is ' ,p0)

p0 is  0.19601837672281777


We find, the average number of cars in the service station is
\begin{align*}
L_q &= \frac{k^k}{k!}p_0 \frac{\rho^{k+1}}{(1-\rho)^2} \\
&\\
L_s&= L_q + k - \sum_{n=0}^{k-1}(k-n)p_n
\end{align*}

We will now have python print out this computation:

In [67]:
Lq = (k**k/(math.factorial(k)))*p0*(rho**(k+1)/((1-rho)**2))*(1-((M-k)*(1-rho)+1)*(rho**(M-k)))

def p(n):
    if n <k:
        return ((k**n)/(math.factorial(n)))*p0*(rho**n)
    elif n>=k and n<=M:
        return ((k**k)/(math.factorial(k)))*p0*(rho**M)
    else:
        return 0

summ = 0.0

for i in range(0,k):
    summ = summ + (k-i)*(p(i))
    
Ls = Lq + k - summ

print('Ls is ',Ls, " is the average number of cars in system")


Ls is  1.7274119448698315  is the average number of cars in system


The average number of cars waiting to be serviced is given by $L_q$ which we computed above.

In [68]:
print('Lq is  ', Lq, '( the avg number of cars in queue)')

Lq is   0.4134762633996937 ( the avg number of cars in queue)


The average amount of time a car spends in the system is given by 
$$
W_s = \frac{Ls}{\lambda}
$$

Again, since we have already computed the values of $L_s$ we have python print the answer:

In [71]:
Ws = (Ls/lamb)
print('Ws equals ',  Ws , 'the average time in days')

Ws equals  0.1439509954058193 the average time in days


## 15.23 Tax consulting firm problem. 

Note: we will be changing names of parameters this problem to math the text, 'Probabilistic Models'. 
First, we recognize that this problem follows a  __M/M/4/FIFO/$\infty$/$\infty$__ model. We first state the given parameters:
\begin{align*}
\lambda &= \frac{10 person}{hour} \\
\mu &= \frac{3 person}{hour}\\
k &= 4 \hspace{5mm} \text{servers} \\
\end{align*}

We will first compute $p_0, p_n, L_s,L_q,W_s $ and $W_q$ since these will be used to answer the questions. 

In [95]:
import math
def fac(s):
    return math.factorial(s)    # too many letters  D:

lamb = 10
mu = 3
s = 4
rho=(lamb/(s*mu))

p0 = 0.0
denom = ((s**s)/math.factorial(s))*((rho*s)/(1-rho))

for n in range(0,s):
    denom = denom+ ((s**n)/(math.factorial(s)))*(rho**n)

p0 = 1/denom

def p(n):
    if n <s:
        return ((s**n)/(math.factorial(n)))*p0*(rho**n)
    elif n>=s:
        return ((s**s)/(math.factorial(s)))*p0*(rho**n)

Lq = ((s**s)/fac(s))*p0*((rho**(s+1))/((1-rho)**2))

summ = 0.0
for n in range(0,s):
    summ = summ+ (s-n)*p(n)

Ls = Lq + s -summ

Ws = Ls/lamb

Wq = Lq/lamb

print('Ls',Ls,'Lq',Lq,'Ws',Ws,'Wq',Wq)

Ls 4.570884381019211 Lq 0.7160400123158885 Ws 0.4570884381019211 Wq 0.07160400123158886


Now that we have successfully computed the variables, we know that the average number of people in the system is 4.57. 

The average number of customers waiting to be served is .71 customers.

The average time a customer spends in the system is .457 hours. 

The average waiting time per customer is .072 hours. 

The average time a tax person spends advisor clients each day is given by 
the expected idle time per server. 

The probability that a customer has to wait is 1-p0 -p1-p2-p3-p4

In [102]:
####idle time per server
idle = 0.0
for n in range(0,s):
    idle = idle+(s-n)*p(n)

idle = (100/s)*idle
print(idle)

3.6288907824169216


In [98]:
print('probability of waiting is ', 1-p0-p(1)-p(2)-p(3)-p(4))

probability of waiting is  0.9016065551076328


Thus the number of hour an advisor works per week is (8-3.6)*5 =22 hours per week

In [104]:
#  expected number of idle workers is
print('expected number of idle workers is ', 4*p0+3*p(1)+2*p(2)+1*p(3) )

expected number of idle workers is  0.14515563129667686
